transformer_engine/pytorch/module/linear.py

```diff
@@ -797,6 +784,63 @@ def __init__(
         else:
             self.gemm_bias_unfused_add = False
 
+        # Clean up weight and bias buffers
+        if self.parameters_split is None:
+            del self.weight_tensor
+            if self.use_bias:
+                del self.bias_tensor
+
+    def init_weight_tensor(self, weights: torch.Tensor) -> torch.Tensor:
+        """
+        Initialize values for the given weight tensor if tensor lives on a real device.
+        """
+        if weights.device == torch.device('meta'):
+            return weights
+
+        # TODO(ksivaman): This functionality works with FP8 outside TE.
+        initialize_affine_weight_gpu(
+            weights,
+            self.param_init_fn,
+            self.get_rng_state_tracker,
+            partition_dim=1 if self.parallel_mode == "row" else 0,
+            stride=1,
+            set_tp_attributes=False,
+        )
+
+        if self.primary_weights_in_fp8:
+            self.init_fp8_metadata()
+            self.fp8_meta["update_amax_and_scale_fwd"] = True
+            weights = Float8Tensor(
+                data=weights,
+                fp8_meta=self.fp8_meta,
+                fp8_meta_index=tex.FP8FwdTensors.GEMM1_WEIGHT,
+            )
+
+        return weights
+
+    def reset_parameters(self, defer_init: bool = False) -> None:
+        """
+        Reset all module parameters to initial values. Unless deferred initialization
+        is specified, all parameters on a 'meta' device are also materialized on a real cuda
+        device before the values are reset to initial.
+        """
+        if defer_init:
+            return
+
+        # materialize all parameters (move to real device, initialize values)
+        for wname, bname in zip(self.weight_names, self.bias_names):
+            if self._parameters[wname].device == torch.device('meta'):
+                self._parameters[wname].to(device='cuda')
```

Comment:
Even though we are mostly concerned with cuda device here, hardcoding this to cuda does not actually feel correct... Not 100% sure how to change it though

Reply:
Paramater reset is doing a cuda cast here only if the tensor is on a meta device. The intent is that the cuda cast can be circumvented as long as the module is cast to a non-meta device *after* initialization, but before the parameter reset.

For example:
```
model = te.Linear(hidden_size, ffn_hidden_size, device='meta')
model.to(device='cpu')
model = torch.distributed.fsdp.FullyShardedDataParallel(model)
```

Unless I'm mistaken, `model.to(device='cpu')` will not trigger memory allocation because the weight/bias tensors are crated as `torch.empty()` when `device='meta'` during module initialization. Consequently, we get the memory allocation only when FSDP triggers the parameter reset, which will not move the tensor to the cuda device because tensor device is not meta at that time.

I'm not sure if this is necessarily the right programming model here, but I think it might be better than the two other alternatives I've thought of:
- Initialize with `te.Linear(...., device='meta', deferred_device='cpu')`, and have the parameter reset always cast to the deferred device.
- Initialize with `te.Linear(..., device='cpu', defer_init=True)`, and internally use meta device initialization based on the deferred init flag.

Neither of these really mesh with PyTorch native module behavior/API so I thought they were worse than the current implementation in the PR. Happy to change it though if we decide otherwise.

Reply:
Generally speaking we have the devices hardcoded to cuda for all allocations since that's the only supported execution. This is guarded by checks for all the activations and weights etc. It looks to be the same for this PR.